code/ION4.1.1/ici/doc/man/man3/sdrstring.3
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.32
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "pod3::sdrstring 3"
.TH pod3::sdrstring 3 "2022-05-20" "perl v5.8.8" "ICI library functions"
.SH "NAME"
sdrstring \- Simple Data Recorder string functions
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\&    #include "sdr.h"
.Ve
.PP
.Vb 4
\&    Object sdr_string_create (Sdr sdr, char *from);
\&    Object sdr_string_dup    (Sdr sdr, Object from);
\&    int    sdr_string_length (Sdr sdr, Object string);
\&    int    sdr_string_read   (Sdr sdr, char *into, Object string);
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\s-1SDR\s0 strings are used to record strings of up to 255 \s-1ASCII\s0 characters
in the heap space of an \s-1SDR\s0.  Unlike standard C strings, which are terminated
by a zero byte, \s-1SDR\s0 strings record the length of the string as
part of the string object.
.PP
To store strings longer than 255 characters, use \fIsdr_malloc()\fR and \fIsdr_write()\fR
instead of these functions.
.IP "Object sdr_string_create(Sdr sdr, char *from)" 4
.IX Item "Object sdr_string_create(Sdr sdr, char *from)"
Creates a \*(L"self\-delimited string\*(R" in the heap of the
indicated \s-1SDR\s0, allocating the required space and copying the
indicated content.  \fIfrom\fR must be a standard C
string for which \fIstrlen()\fR must not exceed 255; if
it does, or if insufficient \s-1SDR\s0 space is available, 0
is returned.  Otherwise the address of the newly created \s-1SDR\s0
string object is returned.  To destroy, just use \fIsdr_free()\fR.
.IP "Object sdr_string_dup(Sdr sdr, Object from)" 4
.IX Item "Object sdr_string_dup(Sdr sdr, Object from)"
Creates a duplicate of the \s-1SDR\s0 string whose address is
\&\fIfrom\fR, allocating the required space and copying the
original string's content.  If insufficient \s-1SDR\s0 space is
available, 0 is returned.  Otherwise the address of the newly
created copy of the original \s-1SDR\s0 string object is returned.  To
destroy, use \fIsdr_free()\fR.
.IP "int sdr_string_length(Sdr sdr, Object string)" 4
.IX Item "int sdr_string_length(Sdr sdr, Object string)"
Returns the length of the indicated self-delimited string (as would
be returned by \fIstrlen()\fR), or \-1 on any error.
.IP "int sdr_string_read(Sdr sdr, char *into, Object string)" 4
.IX Item "int sdr_string_read(Sdr sdr, char *into, Object string)"
Retrieves the content of the indicated self-delimited string into
memory as a standard C string (\s-1NULL\s0 terminated).  Length of \fIinto\fR
should normally be \s-1SDRSTRING_BUFSZ\s0 (i.e., 256) to allow for the largest
possible \s-1SDR\s0 string (255 characters) plus the terminating \s-1NULL\s0.  Returns
length of string (as would be returned by \fIstrlen()\fR), or \-1 on any error.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIsdr\fR\|(3), \fIsdrlist\fR\|(3), \fIsdrtable\fR\|(3), \fIstring\fR\|(3)


code/ION4.1.1/ici/sdr/sdrstring.c
/*
 *	sdrstring.c:	simple data recorder string management
 *			library.
 *
 *	Copyright (c) 2001-2007, California Institute of Technology.
 *	ALL RIGHTS RESERVED.  U.S. Government Sponsorship
 *	acknowledged.
 *
 *	Author: Scott Burleigh, JPL
 *
 *	This library implements the Simple Data Recorder system's
 *	self-delimiting strings.
 *
 *	Modification History:
 *	Date	  Who	What
 *	4-3-96	  APS	Abstracted IPC services and task control.
 *	5-1-96	  APS	Ported to sparc-sunos4.
 *	12-20-00  SCB	Revised for sparc-sunos5.
 *	6-8-07    SCB	Divided sdr.c library into separable components.
 */

#include "sdrP.h"
#include "sdrstring.h"

/*		Private definition of SDR string structure.		*/

typedef unsigned char	SdrStringBuffer[SDRSTRING_BUFSZ];
			/*	SdrString is a low-overhead string
				representation specially designed for
				efficient storage of small strings in
				the SDR.  To avoid time-consuming SDR
				I/O to find the end of a string when
				retrieving it from the SDR, we do not
				NULL-terminate SdrStrings.  Instead
				the first character contains the length
				of the string, which cannot exceed 255.	*/

/*	*	*	String management functions	*	*	*/

Object	Sdr_string_create(const char *file, int line, Sdr sdrv, char *from)
{
	size_t		length = 0;
	Object		string;
	SdrStringBuffer stringBuf;

	if (!sdr_in_xn(sdrv))
	{
		oK(_iEnd(file, line, _notInXnMsg()));
		return 0;
	}

	joinTrace(sdrv, file, line);
	if (from == NULL || (length = strlen(from)) > 255)
	{
		oK(_xniEnd(file, line, from, sdrv));
		return 0;
	}

	string = _sdrmalloc(sdrv, length + 1);
	if (string == 0)
	{
		oK(_iEnd(file, line, from));
		return 0;
	}

	stringBuf[0] = length;
	memcpy(stringBuf + 1, from, length);
	_sdrput(file, line, sdrv, (Address) string, (char *) stringBuf,
			length + 1, SystemPut);
	return string;
}

Object	Sdr_string_dup(const char *file, int line, Sdr sdrv, Object from)
{
	Address		addr = (Address) from;
	unsigned char	length;
	SdrStringBuffer stringBuf;
	Object		string;

	if (!sdr_in_xn(sdrv))
	{
		oK(_iEnd(file, line, _notInXnMsg()));
		return 0;
	}

	joinTrace(sdrv, file, line);
	if (from == 0)
	{
		oK(_xniEnd(file, line, "from", sdrv));
		return 0;
	}

	sdrFetch(length, addr);
	_sdrfetch(sdrv, (char *) (stringBuf + 1), addr + 1, length);
	string = _sdrmalloc(sdrv, length + 1);
	if (string == 0)
	{
		oK(_iEnd(file, line, (char *) (stringBuf + 1)));
		return 0;
	}

	stringBuf[0] = length;
	_sdrput(file, line, sdrv, (Address) string, (char *) stringBuf,
			length + 1, SystemPut);
	return string;
}

int	sdr_string_length(Sdr sdrv, Object string)
{
	unsigned char	length;

	CHKERR(string);
	sdrFetch(length, (Address) string);
	return length;
}

int	sdr_string_read(Sdr sdrv, char *into, Object string)
{
	Address		addr = (Address) string;
	unsigned char	length;

	CHKERR(sdrFetchSafe(sdrv));
	CHKERR(into);
	CHKERR(string);
	sdrFetch(length, addr);
	_sdrfetch(sdrv, into, addr + 1, length);
	*(into + length) = '\0';
	return length;
}
